### PR TITLE
redact `apollo_private` fields

### PR DESCRIPTION
**Pull request by garypen**
The JSON bit is a bit finicky but so is the code in the `tracing_subscriber` crate which I've mainly just lifted here.

fixes: #1862
fixes: #1869

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 1]
```diff
@@ -432,9 +441,22 @@ impl Telemetry {
...
 
-                let sub_builder = tracing_subscriber::fmt::fmt().with_env_filter(
-                    EnvFilter::try_new(log_level).context("could not parse log configuration")?,
-                );
+                let formatter = debug_fn(|writer, field, value| {
+                    if field.name().starts_with("apollo_private") {
+                        // XXX: Are these being correctly delimited?
+                        write!(writer, "{}: null", field)
+                    } else {
+                        write!(writer, "{}: {:?}", field, value)
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/mod.rs:449]: Looks great, just a nit, in the previous formatted version we had fields displayed like this `myfield=content myotherfield=content2` should we try to copy that behavior or not ? Also fields name are in italic but that’s not very important in that case I think

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 2]
```diff
@@ -1232,6 +1258,90 @@ impl ApolloFtv1Handler {
...
+        fields.record(&mut v);
+        v.finish()?;
+        // XXX I can't merge like in the tracing crate. Is this correct?
+        current.fields.pop();
+        new.remove(0);
+        current.fields = format!("{}, {}", current.fields, new);
+        let mut map: BTreeMap<&'_ str, serde_json::Value> =
+            serde_json::from_str(current).map_err(|_| fmt::Error)?;
+        for (key, value) in map.iter_mut() {
+            if key.starts_with("apollo_private") {
```
Comment (BrynCooke) [inline apollo-router/src/plugins/telemetry/mod.rs:1333]: I think it's `apollo.private` not with an underscore.

Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/mod.rs:1333]: Tell a lie, it's not.

**Review by Geal — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 3]
```diff
@@ -432,9 +441,26 @@ impl Telemetry {
                     .map(|s| s.as_str())
                     .unwrap_or("info");
 
-                let sub_builder = tracing_subscriber::fmt::fmt().with_env_filter(
-                    EnvFilter::try_new(log_level).context("could not parse log configuration")?,
-                );
+                let formatter = debug_fn(|writer, field, value| {
+                    if field.name().starts_with("apollo_private")
```
Comment (BrynCooke) [inline apollo-router/src/plugins/telemetry/mod.rs:445]: nit: could suffix field names with `.`

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/mod.rs (hunk 3 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/mod.rs]: done: b8bbb1a2bf768d3e1ed3f47599d2e32410af9224

**Comment by BrynCooke** [comment]
Perf test please

**Pull request merged**